test/sys/test_ptrace.rs

```diff
@@ -197,15 +197,36 @@ fn test_ptrace_syscall() {
             #[cfg(target_arch = "x86")]
             let get_syscall_id = || ptrace::getregs(child).unwrap().orig_eax as libc::c_long;
 
+            // this duplicates `get_syscall_id` for the purpose of testing `ptrace::read_user`.
+            #[cfg(target_arch = "x86_64")]
+            let get_rax_offset = |user_struct_ptr: *const libc::user| {
+                unsafe { &(*user_struct_ptr).regs.orig_rax as *const _ }
+            };
+            #[cfg(target_arch = "x86")]
+            let get_rax_offset = |user_struct_ptr: *const libc::user| {
+                unsafe { &(*user_struct_ptr).regs.orig_eax as *const _ }
+            };
+
+            let get_syscall_from_user_area = || {
+                // Find the offset of `user.regs.rax` (or `eax` for x86)
+                let user_struct = std::mem::MaybeUninit::<libc::user>::uninit();
+                let user_struct_ptr = user_struct.as_ptr();
+                let rax_offset = get_rax_offset(user_struct_ptr) as usize - user_struct_ptr as usize;
```

Comment:
Nix already depends on the `memoffset` crate conditionally, and we can use it in our tests (may need to update Cargo.toml to do so).

It should be possible to use [`memoffset::offset_of`](https://docs.rs/memoffset/latest/memoffset/macro.offset_of.html) to simplify this.

Reply:
Great suggestion, that should be done :+1: